File: src-lib/Language/Mill/Parse.hs
module Language.Mill.Parse where

import Control.Applicative ((<$>), (<|>), (<*), (<*>), (*>))
import Control.Monad (foldM)
import Language.Mill.AST
import Language.Mill.AST.ID (newID)
import Language.Mill.Lex
import Language.Mill.Module (ModuleName(..))
import Text.Parsec (eof, sepBy, sepBy1, try, sepEndBy, many, optionMaybe)

module_ :: Parser Module
module_ = Module <$> many decl <* eof

name :: Parser Name
name = do
    part1 <- identifier
    part2 <- optionMaybe (dot *> identifier)
    return $ case part2 of
        Nothing -> UnqualifiedName part1
        Just id -> QualifiedName part1 id

type_ :: Parser Type
type_ = namedType <|> try subType <|> tupleType

namedType :: Parser Type
namedType = NamedType <$> newID <*> name

subType :: Parser Type
subType = do
    id <- newID
    parameterTypes <- (openingParenthesis *> type_ `sepEndBy` comma <* closingParenthesis)
    fatArrow
    returnType <- type_
    return $ SubType id parameterTypes returnType

tupleType :: Parser Type
tupleType = TupleType <$> newID <*> (openingParenthesis *> type_ `sepEndBy` comma <* closingParenthesis)

parameter :: Parser Parameter
parameter = do
    id <- newID
    name <- identifier
    colon
    paramType <- type_
    return $ Parameter id name paramType

parameterList :: Parser ParameterList
parameterList = openingParenthesis *> parameter `sepEndBy` comma <* closingParenthesis

expr :: Parser Expr
expr = callExpr

callExpr :: Parser Expr
callExpr = do
    callee <- primaryExpr
    argumentLists <- many argumentList
    foldM (\a b -> (\id -> CallExpr id a b) <$> newID) callee argumentLists
    where
        argumentList :: Parser [Expr]
        argumentList = openingParenthesis *> expr `sepEndBy` comma <* closingParenthesis

primaryExpr :: Parser Expr
primaryExpr = try structLiteralExpr <|> nameExpr <|> stringLiteralExpr <|> blockExpr

nameExpr :: Parser Expr
nameExpr = NameExpr <$> newID <*> name

stringLiteralExpr :: Parser Expr
stringLiteralExpr = StringLiteralExpr <$> newID <*> stringLiteral

blockExpr :: Parser Expr
blockExpr = BlockExpr <$> newID <*> (openingBrace *> many stmt <* closingBrace)

structLiteralExpr :: Parser Expr
structLiteralExpr = StructLiteralExpr <$> type_ <*> (openingBrace *> fieldValue `sepBy` comma <* closingBrace)
    where
        fieldValue :: Parser FieldValue
        fieldValue = FieldValue <$> identifier <*> (colon *> expr)

stmt :: Parser Stmt
stmt = (ExprStmt <$> expr) <|> (DeclStmt <$> decl)

decl :: Parser Decl
decl = aliasDecl <|> importDecl <|> try structDecl <|> subDecl <|> foreignSubDecl

aliasDecl :: Parser Decl
aliasDecl = do
    id <- newID
    aliasKeyword
    name <- identifier
    equalsSign
    original <- type_
    return $ AliasDecl id name original

importDecl :: Parser Decl
importDecl = do
    id <- newID
    moduleName <- ModuleName <$> (importKeyword *> identifier `sepBy1` dot)
    return $ ImportDecl id moduleName

structDecl :: Parser Decl
structDecl = do
    id <- newID
    structKeyword
    name <- identifier
    openingBrace
    fields <- many field
    closingBrace
    return $ StructDecl id name fields
    where
        field :: Parser Field
        field = Field <$> identifier <*> (colon *> type_)

subDecl :: Parser Decl
subDecl = do
    id <- newID
    subKeyword
    name <- identifier
    params <- parameterList
    colon
    retType <- type_
    body <- blockExpr
    return $ SubDecl id name params retType body

foreignSubDecl :: Parser Decl
foreignSubDecl = do
    id <- newID
    foreignKeyword
    library <- ForeignLibrary <$> stringLiteral
    subKeyword
    cconv <- CallingConvention <$> name
    name <- identifier
    params <- parameterList
    colon
    retType <- type_
    return $ ForeignSubDecl id library cconv name params retType

File: src/main/scala/org/vlinderlang/vlinderc/parse/parse.scala
package org.vlinderlang.vlinderc.parse

import org.vlinderlang.vlinderc.ast._
import org.vlinderlang.vlinderc.ModuleName
import scala.util.parsing.combinator.Parsers

private[parse] object Parser extends Parsers {
  override type Elem = Token

  private def identifier: Parser[String] =
    accept("identifier", { case Identifier(name) => name })

  private def stringLiteral: Parser[String] =
    accept("string literal", { case StringLiteral(value) => value })

  def name: Parser[Name] = Vector(
    for { m <- identifier; _ <- Period; n <- identifier } yield QualifiedName(m, n),
    identifier ^^ UnqualifiedName
  ).reduce(_ | _)

  def decl: Parser[Decl] = Vector(
    aliasDecl,
    importDecl,
    structDecl,
    subDecl,
    unionDecl
  ).reduce(_ | _)

  def aliasDecl: Parser[Decl] = for {
    _ <- Typealias
    name <- identifier
    _ <- Eq
    underlyingType <- typeExpr
    _ <- Semicolon
  } yield AliasDecl(name, underlyingType)

  def importDecl: Parser[Decl] = {
    def moduleName = rep1sep(identifier, Period) ^^ (ModuleName(_: _*))
    Import ~> moduleName <~ Semicolon ^^ ImportDecl
  }

  def structDecl: Parser[Decl] = for {
    _ <- Struct
    name <- identifier
    _ <- LeftBrace
    fields <- (for {
      n <- identifier
      _ <- Colon
      t <- typeExpr
      _ <- Semicolon
    } yield (n, t)).* ^^ (_.toVector)
    _ <- RightBrace
    _ <- Semicolon
  } yield StructDecl(name, fields)

  def subDecl: Parser[Decl] = {
    def param: Parser[(String, TypeExpr)] = for {
      name <- identifier
      _ <- Colon
      type_ <- typeExpr
    } yield (name, type_)
    for {
      _ <- Sub
      name <- identifier
      params <- LeftParen ~> repsep(param, Comma) <~ RightParen ^^ (_.toVector)
      _ <- Colon
      returnType <- typeExpr
      body <- blockExpr
      _ <- Semicolon
    } yield SubDecl(name, params, returnType, body)
  }

  def unionDecl: Parser[Decl] = for {
    _ <- Union
    name <- identifier
    _ <- LeftBrace
    constructors <- (identifier <~ Semicolon ^^ { i => (i, Vector()) }).* ^^ (_.toVector)
    _ <- RightBrace
    _ <- Semicolon
  } yield UnionDecl(name, constructors)

  def expr: Parser[Expr] = callExpr

  def callExpr: Parser[Expr] = {
    val suffix = LeftParen ~> repsep(expr, Comma) <~ RightParen ^^ (_.toVector)
    for {
      callee <- primaryExpr
      suffixes <- suffix.*
    } yield suffixes.foldLeft(callee) { (callee, suffix) => CallExpr(callee, suffix) }
  }

  def primaryExpr: Parser[Expr] = Vector(
    structLiteralExpr,
    nameExpr,
    blockExpr,
    booleanLiteralExpr,
    stringLiteralExpr
  ).reduce(_ | _)

  def nameExpr: Parser[Expr] =
    name ^^ NameExpr

  def blockExpr: Parser[Expr] =
    LeftBrace ~> (expr <~ Semicolon).* <~ RightBrace ^^ (BlockExpr(_: _*))

  def booleanLiteralExpr: Parser[Expr] =
    (True ^^^ true | False ^^^ false) ^^ BooleanLiteralExpr

  def stringLiteralExpr: Parser[Expr] =
    stringLiteral ^^ StringLiteralExpr

  def structLiteralExpr: Parser[Expr] = {
    def field: Parser[(String, Expr)] = for {
      name <- identifier
      _ <- Colon
      value <- expr
    } yield (name, value)
    for {
      struct <- typeExpr
      fields <- LeftBrace ~> repsep(field, Comma) <~ RightBrace ^^ (_.toVector)
    } yield StructLiteralExpr(struct, fields)
  }

  def typeExpr: Parser[TypeExpr] = Vector(
    nameTypeExpr,
    subTypeExpr,
    tupleTypeExpr
  ).reduce(_ | _)

  def nameTypeExpr: Parser[TypeExpr] =
    name ^^ NameTypeExpr

  def tupleTypeExpr: Parser[TypeExpr] =
    LeftParen ~> repsep(typeExpr, Comma) <~ RightParen ^^ (TupleTypeExpr(_: _*))

  def subTypeExpr: Parser[TypeExpr] = for {
    paramTypes <- LeftParen ~> repsep(typeExpr, Comma) <~ RightParen ^^ (_.toVector)
    _ <- EqGT
    returnType <- typeExpr
  } yield SubTypeExpr(paramTypes, returnType)

  def module(name: ModuleName): Parser[Module] =
    decl.* ^^ { decls => Module(name, decls.toVector) }
}
